crates/svm-storage/src/kv/ffi.rs

```diff
@@ -1,80 +1,129 @@
-//         let key_len = key.len() as u32;
-
-//         let mut value_len = 0;
-
-//         let value_ptr = unsafe { (self.get_fn)(key_ptr, key_len, &mut value_len) };
-
-//         if value_len > 0 {
-//             let value = unsafe { std::slice::from_raw_parts(value_ptr, value_len as usize) };
-
-//             Some(value.to_vec())
-//         } else {
-//             None
-//         }
-//     }
-
-//     fn store(&mut self, changes: &[(&[u8], &[u8])]) {
-//         for (k, v) in changes.iter() {
-//             let key_ptr = k.as_ptr();
-//             let val_ptr = v.as_ptr();
-
-//             let key_len = k.len() as u32;
-//             let val_len = v.len() as u32;
-
-//             unsafe {
-//                 (self.set_fn)(key_ptr, key_len, val_ptr, val_len);
-//             }
-//         }
-
-//         unsafe {
-//             (self.commit_fn)();
-//         }
-//     }
-// }
-
-// impl StatefulKVStore for ExternKV {
-//     fn rewind(&mut self, state: &State) {
-//         let length = State::len() as u32;
-//         let ptr = state.as_ptr();
-
-//         unsafe {
-//             (self.rewind_fn)(ptr, length);
-//         }
-//     }
-
-//     #[must_use]
-//     fn head(&self) -> State {
-//         let mut length = 0u32;
-//         unsafe {
-//             let ptr = (self.head_fn)(&mut length);
-
-//             assert_eq!(length, State::len() as u32);
-
-//             State::from(ptr)
-//         }
-//     }
-// }
+use super::StatefulKV;
+
+use svm_common::State;
+
+const BUF_SIZE: usize = 1024;
+static mut BUF: [u8; BUF_SIZE] = [0; BUF_SIZE];
+
+/// # Get a Key's Value
+///
+/// Gets a key's matching value.
+/// The value buffer to copy the value to is allocated by `SVM`.
+/// The buffer maximum size is 1024 bytes. (This value should be more than enough).
+///
+/// * key_ptr   - a raw pointer to the key's first byte
+/// * key_len   - key's byte-length
+/// * value_ptr - a raw pointer to the value's buffer first byte
+/// * value_len - a pointer
+pub type GetFn = unsafe extern "C" fn(*const u8, u32, *mut u8, *mut u32);
+
+/// # Sets a Key's Value
+///
+/// Sets a value for a key.
+///
+/// * key_ptr   - a raw pointer to the key's first byte
+/// * key_len   - key's byte-length
+/// * value_ptr - a raw pointer to the value's first byte
+/// * value_len - a raw pointer to the value's first byte
+pub type SetFn = unsafe extern "C" fn(*const u8, u32, *const u8, u32);
+
+/// # Discard
+///
+/// Notifies the `Host` to discard the current executed transaction changes.
+pub type DiscardFn = unsafe extern "C" fn();
+
+/// # Checkpoint
+///
+/// Notifies the `Host` to checkpoint key-value `State`.
+/// Returns the `State` derived for the checkpoint.
+///
+/// Computing a checkpoint doesn't guarantee that the pendind changes and checkpoint
+/// have been persisted. It's up to the `Host` to determine when to save data for long-term usage.
+pub type CheckpointFn = unsafe extern "C" fn(*mut u8);
```

Comment:
Is this the same as `commit`? if so, why using this term? (checkpoint/discard instead commit/rollback)

I thought that checkpoint is about rewind-related optimization.

Reply:
I've decided not to use the term `commit` here since I wanted to stress that it's up to the `Host` to decide when
to do the `committing/flushing` to disk.

So `checkpoint` here means: 
```
"I'm SVM has completed my part in executing the transaction.
I'd signaling the Host that I'm done and I'd like to know my new State.
It's up to you the Host, to determine when you want to persist the changes."
```